### ISSUE TITLE
Resque integration causes <ThreadErrors> when sending signals to workers

### ISSUE THREAD

**Issue by matthewhively** [newrelic/newrelic-ruby-agent#1020]
## Description
It appears that when the Newrelic agent is attached to Resque worker processes that something about the logging code causes resque workers to no longer be thread safe and thus throw these errors.
`ThreadError: can't be called from trap context` whenever workers trap some signals. 
The Resque explicitly uses MonoLogger as its logger  to prevent these sorts of threading issues. I'm not sure whether Newrelic is replacing the logger, or monkey-patching thread-unsafe methods back into it.
I was able to blacklist the `resque:work` rake task through the newrelic.yml file to stop these exceptions.

## Expected Behavior
I would expect that Resque worker threads can log all of their messages without raising exceptions.

## [Troubleshooting]
Relevant portion of Resque log:
```
D, [2022-03-09T19:01:50.972179 #36073] DEBUG -- : Checking test
D, [2022-03-09T19:01:50.974433 #36073] DEBUG -- : Found job on test
D, [2022-03-09T19:01:50.975878 #36073] DEBUG -- : resque-2.2.0: Processing test since 1646881310 [TestJob]
I, [2022-03-09T19:01:50.976226 #36073]  INFO -- : got: (Job{test} | TestJob | ["foobar08.txt", "something in 8 seconds", 8])
I, [2022-03-09T19:01:50.976412 #36073]  INFO -- : Running before_fork hooks with [(Job{test} | TestJob | ["foobar08.txt", "something in 8 seconds", 8])]
D, [2022-03-09T19:01:50.980398 #36073] DEBUG -- : resque-2.2.0: Forked 36111 at 1646881310
I, [2022-03-09T19:01:50.986422 #36111]  INFO -- : Running after_fork hooks with [(Job{test} | TestJob | ["foobar08.txt", "something in 8 seconds", 8])]
E, [2022-03-09T19:01:56.538140 #36073] ERROR -- : Failed to start worker : #<ThreadError: can't be called from trap context>
I, [2022-03-09T19:01:59.113721 #36111]  INFO -- : done: (Job{test} | TestJob | ["foobar08.txt", "something in 8 seconds", 8])
```

Relevant portion of stack trace:
```
.rvm/gems/ruby-2.5.7/gems/newrelic_rpm-8.1.0/lib/new_relic/agent/stats_engine.rb:24:in `synchronize',
.rvm/gems/ruby-2.5.7/gems/newrelic_rpm-8.1.0/lib/new_relic/agent/stats_engine.rb:24:in `with_stats_lock',
.rvm/gems/ruby-2.5.7/gems/newrelic_rpm-8.1.0/lib/new_relic/agent/stats_engine.rb:74:in `record_unscoped_metrics',
.rvm/gems/ruby-2.5.7/gems/newrelic_rpm-8.1.0/lib/new_relic/agent/stats_engine.rb:65:in `tl_record_unscoped_metrics',
.rvm/gems/ruby-2.5.7/gems/newrelic_rpm-8.1.0/lib/new_relic/agent.rb:224:in `increment_metric',
.rvm/gems/ruby-2.5.7/gems/newrelic_rpm-8.1.0/lib/new_relic/agent/instrumentation/logger/instrumentation.rb:44:in `format_message_with_tracing',
.rvm/gems/ruby-2.5.7/gems/newrelic_rpm-8.1.0/lib/new_relic/agent/instrumentation/logger/prepend.rb:10:in `format_message',
.rvm/rubies/ruby-2.5.7/lib/ruby/2.5.0/logger.rb:472:in `add',
.rvm/rubies/ruby-2.5.7/lib/ruby/2.5.0/logger.rb:527:in `info',
.rvm/gems/ruby-2.5.7/gems/resque-2.2.0/lib/resque/logging.rb:8:in `log',
.rvm/gems/ruby-2.5.7/gems/resque-2.2.0/lib/resque/worker.rb:940:in `log_with_severity',
.rvm/gems/ruby-2.5.7/gems/resque-2.2.0/lib/resque/worker.rb:437:in `shutdown',
.rvm/gems/ruby-2.5.7/gems/resque-2.2.0/lib/resque/worker.rb:400:in `block in register_signal_handlers',
```

## Steps to Reproduce
Add Newrelic gem to a rails project with Resque.
Start a Resque worker.
Have Resque worker processing a job (can be a simple job to sleep for 10 sec)
Send sigTERM or sigQUIT or sigUSR1 to the worker.
Observe the Resque log for a thread error.

## Your Environment
Ruby Version 2.5.7 (through RVM)
Rails 5.2.5
Resque 2.2.0
NewRelic 8.5.0

## Additional context
Discussion about Resque's use of MonoLogger
https://github.com/resque/resque/pull/1729

**Comment by tannalynn** [comment]
Hello @matthewhively 
Thank you for bringing this issue to our attention.

In the stack trace you've provided, I see that it is going through the new Logger instrumentation that was added in 8.1.0, so it's possible that something there is interfering with resque. Disabling this instrumentation may avoid this error and might be a good workaround to use while we are looking into the issue. You can disable this instrumentation by adding the following setting to your newrelic.yml config file
```
instrumentation:
  logger: disabled
```
Hopefully that helps as a workaround for now while we investigate the root cause.

**Comment by matthewhively** [comment]
Thanks, I'll try to use that config instead of excluding the whole rake task and I'll let you know if it sidestepped the issue.

As a quick note the partial stacktrace I included says newrelic 8.1.0, but I got the same result after I updated the gem to 8.5.0

**Comment by matthewhively** [comment]
Yes, I can confirm that setting `instrumentation.logger: disabled` did indeed work around this issue.
We'll be watching this issue for any further developments.

**Comment by tannalynn** [comment]
That's great! I'm glad to hear that it did workaround the issue for you. I've spent some more time looking into it this morning and, with your reproduction instructions, I was able to reproduce this issue with a test app. We are currently working on a branch that reworks and adds on to the logging instrumentation that is currently in the agent, and part of the work that has already been done over  there was related to resque. So I tried switching to that branch of the agent to try to repro the error again, and wasn't seeing it happen anymore. So it seems like we may have a branch already that resolves this issue. 

If you'd like to try the branch out to confirm if it resolves the error for you as well, you could install the agent from that branch. Remove the configuration option I mentioned previously (`instrumentation.logger`) from your newrelic.yml file and then add the agent to your Gemfile using the logging-initiative branch: 
```
gem 'newrelic_rpm', :git => 'https://github.com/newrelic/newrelic-ruby-agent.git',  :branch => 'logging-initiative'
```

Hopefully it does indeed resolve the error for you as well, please let me know if you continue to have any issues on that branch.

**Comment by matthewhively** [comment]
Yes I can confirm that the `logging-initiative` branch currently does not cause resque worker ThreadErrors.

**Comment by tannalynn** [comment]
Today we released version 8.6.0 which contained the fixes for this issue. Thank you again for letting us know about the issue!
